codalab/lib/beam/blobstorageuploader.py

```diff
@@ -0,0 +1,98 @@
+from apache_beam.io.filesystemio import Uploader, UploaderStream
+from azure.storage.blob import (
+    BlobServiceClient,
+    ContentSettings,
+    BlobBlock,
+)
+from apache_beam.io.azure.blobstorageio import parse_azfs_path
+import tempfile
+import io
+import base64
+from codalab.worker.un_gzip_stream import BytesBuffer
+
+class BlobStorageUploader(Uploader):
+  """An improved version of apache_beam.io.azure.blobstorageio.BlobStorageUploader
+  that handles multipart streaming (block-by-block) uploads.
+  TODO (Ashwin): contribute this back upstream to Apache Beam (https://github.com/codalab/codalab-worksheets/issues/3475).
+  """
+  def __init__(self, client, path, mime_type='application/octet-stream'):
+    self._client = client
+    self._path = path
+    self._container, self._blob = parse_azfs_path(path)
+    self._content_settings = ContentSettings(mime_type)
+
+    self._blob_to_upload = self._client.get_blob_client(
+        self._container, self._blob)
+
+    self.block_number = 1
+    self.buffer = BytesBuffer()
+    self.block_list = []
+
+  def put(self, data):
+    # Note that Blob Storage currently can hold a maximum of 100,000 uncommitted blocks.
+    # This means that with this current implementation, we can upload a file with a maximum
+    # size of 10 TiB to Blob Storage. To exceed that limit, we must either increase MIN_WRITE_SIZE
+    # or modify the implementation of this class to call commit_block_list more often (and not
+    # just at the end of the upload). 
+    MIN_WRITE_SIZE = 100 * 1024 * 1024
+    MAX_WRITE_SIZE = 2 * 1024 * 1024 * 1024
+
+    self.buffer.write(data.tobytes())
+
+    while len(self.buffer) >= MIN_WRITE_SIZE:
+      # Take the first chunk off the buffer and write it to Blob Storage
+      chunk = self.buffer.read(MAX_WRITE_SIZE)
+      self._write_to_blob(chunk)
+
+  def _write_to_blob(self, data):
+    # block_id's have to be base-64 strings normalized to have the same length.
+    block_id = base64.b64encode('{0:-32d}'.format(self.block_number).encode()).decode()
+    
+    self._blob_to_upload.stage_block(block_id, data)
+    self.block_list.append(BlobBlock(block_id))
+    self.block_number = self.block_number + 1
```

Comment:
Does azure auto-delete blocks if an upload stops midway?

Reply:
Yes -- looks like it auto-deletes blocks after a week: https://stackoverflow.com/questions/62021724/azure-block-blob-staging-area-block-expiration